### PR TITLE
AP_GPS: add binary CRC message parser abstract class

### PR DESCRIPTION
**Pull request by Femtomes**
1. Add binary CRC message parser abstract class, 
2. Update nova driver use the parser

### CODE REVIEW

File: libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    virtual uint8_t *get_message_body_buff() = 0;
+    bool parse(uint8_t data);
+    virtual bool process_message() = 0;
+
+private:
+    void reset_parse();
+    uint32_t crc32_value(uint32_t icrc);
+    uint32_t calculate_block_crc32(uint32_t length, uint8_t *buffer, uint32_t crc);
+
+    const uint32_t CRC32_POLYNOMIAL = 0xEDB88320L;
```
Comment (amilcarlucas) [inline libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.h:43]: static constexp

File: libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    const std::vector<uint8_t> _preambles;
+    uint8_t _min_header_length;
+    uint8_t _max_header_length;
+    uint16_t _max_body_length;
+
+    uint8_t _decode_step = 0;   // step number of message decode
```
Comment (amilcarlucas) [inline libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.h:50]: No need to init variables with 0. It is done automatically

File: libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.cpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    _decode_step = 0;
+    _read_size = 0;
+    _preamble_step = 0;
+}
+
+uint32_t AP_GPS_BinaryCRCMessageParser::crc32_value(uint32_t icrc)
+{
+    int i;
+    uint32_t crc = icrc;
+    for ( i = 8 ; i > 0; i-- ) {
```
Comment (amilcarlucas) [inline libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.cpp:137]: reduce the scope of the i variable, declare it inside the for loop

File: libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.cpp [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    }
+    return crc;
+}
+
+uint32_t AP_GPS_BinaryCRCMessageParser::calculate_block_crc32(uint32_t length, uint8_t *buffer, uint32_t crc)
+{
+    while ( length-- != 0 ) {
+        crc = ((crc >> 8) & 0x00FFFFFFL) ^ (crc32_value(((uint32_t) crc ^ *buffer++) & 0xff));
+    }
+    return( crc );
```
Comment (amilcarlucas) [inline libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.cpp:152]: why the parentheses?

**Review by amilcarlucas — CHANGES_REQUESTED** [review]
Does this save memory? Will it be reused? What is the use case?

**Review by amilcarlucas — APPROVED** [review]

File: libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    virtual uint8_t *get_message_body_buff() = 0;
+    bool parse(uint8_t data);
+    virtual bool process_message() = 0;
+
+private:
+    void reset_parse();
+    uint32_t crc32_value(uint32_t icrc);
+    uint32_t calculate_block_crc32(uint32_t length, uint8_t *buffer, uint32_t crc);
+
+    static const uint32_t CRC32_POLYNOMIAL = 0xEDB88320L;
```
Comment (amilcarlucas) [inline libraries/AP_GPS/AP_GPS_BinaryCRCMessageParser.h:43]:
```suggestion
    static constexpr uint32_t CRC32_POLYNOMIAL = 0xEDB88320L;
```

File: libraries/AP_GPS/AP_GPS_NOVA.cpp [hunk 6]
```diff
@@ -40,10 +40,9 @@ do {                                            \
 
 AP_GPS_NOVA::AP_GPS_NOVA(AP_GPS &_gps, AP_GPS::GPS_State &_state,
                        AP_HAL::UARTDriver *_port) :
-    AP_GPS_Backend(_gps, _state, _port)
+    AP_GPS_Backend(_gps, _state, _port),
+    AP_GPS_BinaryCRCMessageParser({0xaa,0x44,0x12},4,sizeof(nova_msg.header),sizeof(nova_msg.data))
```
Comment (amilcarlucas) [inline libraries/AP_GPS/AP_GPS_NOVA.cpp:44]:
```suggestion
    AP_GPS_BinaryCRCMessageParser({0xaa,0x44,0x12}, 4, sizeof(nova_msg.header), sizeof(nova_msg.data))
```

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

**Comment by Femtomes** [comment]
Are there any other issues?

**Comment by Vkd4U2NtVkZOVg** [comment]
Hi @Femtomes , I have a question about this crc code. and also I test it on the nova driver, why can we use crc_crc32 from the AP_Math instead of write a new one?

**Comment by Femtomes** [comment]
> Hi @Femtomes , I have a question about this crc code. and also I test it on the nova driver, why can we use crc_crc32 from the AP_Math instead of write a new one?

Oh! I don't know that the common source code of crc32 is already in the project. If it can match Nova's crc, I would prefer to use it.

**Comment by Femtomes** [comment]
Reorganized code at #18690

**Pull request closed without merging**